### DISCUSSION TITLE
Image Crop along with Scale and Rotate

### DISCUSSION THREAD

**Discussion by Pratham95VP** [AvaloniaUI/Avalonia#14041] [General]
I'm relatively new to C# and Avalonia UI, currently working on a personal learning project. I'm attempting to implement image cropping functionality within Avalonia UI, using libraries such as Emgu CV and SkiaSharp. However, I'm facing challenges in achieving basic image editing capabilities similar to those found in Android, iOS, or standard image editors.

I've experimented with pointer events, akin to mouse events in WPF, but haven't been able to produce the desired outcome. My goal is to implement a straightforward image cropping feature that mimics the simplicity of basic editing found in popular image editors.

I would greatly appreciate any guidance or assistance on how to proceed with this task in Avalonia UI.

**Comment by thevortexcloud** [comment]
Avalonia does have `CroppedBitmap` which may work better than you trying to do it from scratch yourself. I am not sure if it supports on the fly cropping though as I have never used it.

In any case, it sounds like you want to use a `Canvas` and then place some sort of [overlay ontop of the image](https://reference.avaloniaui.net/api/Avalonia/Visual/5DED64FE). The overlay can then consume all the input events and draw what ever shapes it needs over the image. For simple cropping it's a matter of calculating the overlay size and offsets and applying it to the source image. Adorners may also help you here with making handles for the cropping overlay. 

 You can also take a look at https://github.com/wieslawsoltes/PanAndZoom which also probably contains a lot of functionality you want.

**Reply by Pratham95VP** [reply]
Thanks for the input. I tried using canvas. The issue is I am not able to free crop like we can do in android or iso or any other editor. 
for example this is the sample code. 
```
 <Canvas  x:Name="Canvas" HorizontalAlignment="Left" VerticalAlignment="Top"  >
     <Image x:Name="Image" Stretch="Fill"Width="480" Height="320" ClipToBounds="True" ZIndex="0" />
<!-- Cropping Rectrangle -->
 <Rectangle x:Name="Rectangle" Stretch="Uniform" HorizontalAlignment="Left" VerticalAlignment="Top" 
	    Stroke="Black" Fill="Transparent" IsVisible="False" StrokeThickness="2"  Width="320"
            Height="200" StrokeLineCap="Round" StrokeDashArray="2,2,0,2" ZIndex="1" />		
```  
 </Canvas>
Here I am not able to drag the edges of the Rectangle. And if the Image is Scaled or rotated then the cropped part is not as it should be

**Reply by timunie** [reply]
We are not an image editor library. You can use RenderTargetBitmap to draw on your own or use any other image lib like ImageMagic to handle the image and return the pixels to the RTB. 

Also you can inherit control and render on your own.

**Comment by Pratham95VP** [comment]
I have found some solution and now I am able to do cropping the way I wanted . Still I need guidance with the image cropping of RotateTransform image.
